File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        curr=head
        stack=[]
        if not head:
            return None
        while curr:
            stack.append(curr)
            curr=curr.next
        head=stack.pop()
        curr=head
        while stack:
            node=stack.pop()
            curr.next=node
            node.next=None
            curr=curr.next
        return head

File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // if(!head)
        //     return NULL;
        // ListNode* prev=NULL;
        // ListNode* curr=head;
        // ListNode* next=curr->next;



        // while(curr){
        //     ListNode* next=curr->next;

        //     curr->next=prev;

        //     prev=curr;
        //     curr=next;
        // }
        // return prev; 
        if (!head)
            return nullptr;
        if (!head->next)
            return head;
        ListNode* reversedListLast=head->next;
        ListNode* reversed=reverseList(head->next);
        reversedListLast->next=head;
        head->next=nullptr;
        return reversed;
    }
};